### DISCUSSION TITLE
Service account's token and CA cert still auto-mounted into pods by default

### DISCUSSION THREAD

**Discussion by IceManGreen** [k3s-io/k3s#7367] [Q&A] [answered]
**Environmental Info:**
K3s Version: 
1.26.3+k3s1

Node(s) CPU architecture, OS, and Version: 
Linux my-hostname 5.10.0-19-amd64 #1 SMP Debian 5.10.149-2 (2022-10-21) x86_64 GNU/Linux
(Debian 11)

Cluster Configuration:
1 server, 2 agents

**Describe the bug:**
When a Pod is created, the service acccount's token is still auto-mounted into `/var/run/secrets/kubernetes.io/serviceaccount/token`.
Same story for the CA cert in `/var/run/secrets/kubernetes.io/serviceaccount/ca.crt`.
**However**, this behavior should not happen since Kubernetes v1.24 (beta) and Kubernetes 1.26 (GA LegacyServiceAccountTokenNoAutoGeneration).

**Steps To Reproduce:**
- K3S server install command : 
`curl -sfL https://get.k3s.io | INSTALL_K3S_EXEC="server --disable traefik" K3S_TOKEN="mytoken" sh -s`

- K3S agents install command : 
```sh
# agent 1 on node 1
curl -sfL https://get.k3s.io | INSTALL_K3S_EXEC="agent" K3S_TOKEN="mytoken" K3S_NODE_NAME="agent1" sh -s - --server https://master:6443
# agent 2 on node 2
curl -sfL https://get.k3s.io | INSTALL_K3S_EXEC="agent" K3S_TOKEN="mytoken" K3S_NODE_NAME="agent2" sh -s - --server https://master:6443
```

- Create any pod running a shell like `kubectl run -it --image alpine bash`.

- Check the token content with `cat /var/run/secrets/kubernetes.io/serviceaccount/token`. Same for `ca.crt`

**Expected behavior:**
No token and ca.crt files existing in the pod.

**Actual behavior:**
Token and ca.crt files in pod.

**Additional context / logs:**
@Typhlos has worked on it with me

More info about the GA:
https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.26.md#feature-4
https://github.com/kubernetes/enhancements/tree/master/keps/sig-auth/2799-reduction-of-secret-based-service-account-token#legacyserviceaccounttokennoautogeneration

**Comment by brandond** [comment, marked as answer]
I think you might be confused. As per the KEP:
> LegacyServiceAccountTokenNoAutoGeneration:
> Token Controller stops auto-creating secret for service accounts.

The FeatureGate you linked is for no longer auto-creating **Secrets** for ServiceAccounts. It does not change anything about how **tokens** are created, or whether or not the default ServiceAccount is mounted into pods by default.

The **token** you are seeing does not come from a Secret, it comes from the [bound service account token volume](https://kubernetes.io/docs/reference/access-authn-authz/service-accounts-admin/#bound-service-account-token-volume).

Did you perhaps want to modify the namespace's default ServiceAccount to disable automounting by default?
https://kubernetes.io/docs/tasks/configure-pod-container/configure-service-account/#opt-out-of-api-credential-automounting

**Comment by Typhlos** [comment]
Indeed, we were confused, we thought that this FeatureGate disabled automounting in the pod but it does not.

We think that the token should not be mounted by default inside of the pods as it gives an authenticated access to the API and could allow an attacker to leverage that access to the API after compromising a pod. But it seems that there was already such a discussion in the kubernetes project and that change was refused due to backward compatibility (see https://github.com/kubernetes/kubernetes/issues/57601 ).

Thank you for your answer!

**Reply by brandond** [reply]
You can turn it off on a per-SA level if you don't want it mounted.
